src/java/org/apache/cassandra/metrics/AccordMetrics.java

```diff
@@ -0,0 +1,279 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.metrics;
+
+import java.lang.reflect.Field;
+import java.util.concurrent.TimeUnit;
+
+import accord.api.EventsListener;
+import accord.local.Command;
+import accord.primitives.Deps;
+import accord.primitives.PartialDeps;
+import accord.primitives.Timestamp;
+import accord.primitives.TxnId;
+import com.codahale.metrics.Counting;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Timer;
+import org.apache.cassandra.service.accord.AccordService;
+
+import static java.util.concurrent.TimeUnit.MICROSECONDS;
+import static org.apache.cassandra.metrics.CassandraMetricsRegistry.Metrics;
+
+public class AccordMetrics
+{
+    public final static AccordMetrics readMetrics = new AccordMetrics("ro");
+    public final static AccordMetrics writeMetrics = new AccordMetrics("rw");
+
+    /**
+     * The time between start on the coordinator and commit on this replica.
+     */
+    public final Timer commitLatency;
+
+    /**
+     * The time between start on the coordinator and execution on this replica.
+     */
+    public final Timer executeLatency;
+
+    /**
+     * The time between start on the coordinator and application on this replica.
+     */
+    public final Timer applyLatency;
+
+    /**
+     * Duration of applying changes.
+     */
+    public final Timer applyDuration;
+
+    /**
+     * The number of dependencies of the transaction on this replica.
+     */
+    public final Histogram deps;
+
+    public final Meter progressLogSize;
+
+    /**
+     * The number of dependencies of the transaction.
+     */
+    public final Histogram coordinatorDeps;
+
+    /**
+     * The number of fast path transactions executed on this coordinator.
+     */
+    public final Meter coordinatorFastPathTrx;
+
+    /**
+     * The number of slow path transactions executed on this coordinator.
+     */
+    public final Meter coordinatorSlowPathTrx;
+
+    /**
+     * The number of preempted transactions on this coordinator.
+     */
+    public final Meter coordinatorPreemptedTrx;
+
+    /**
+     * The number of timed out transactions on this coordinator.
+     */
+    public final Meter coordinatorTimedoutTrx;
```

Comment:
```suggestion
    public final Meter timeouts;
```
Similar to what we have in `ClientRequestMetrics`?

Reply:
Wait, after looking at this again, what is the difference between the timeout and preempted metrics here and the ones in `AccordClientRequestMetrics`?

Reply:
Indeed. I was going to ask about that - `ACRM` are tied to the statement processing initiated by the client, on the coordinator hit by the client request. What about the events at the recovery coordinator?

Reply:
Yeah, it seems like the overlap is only partial? Trying to remember the details of the protocol, but the hooks we've put in `CoordinatePreAccept` would only apply to transactions initiated by the client, and would be the double-counted items (well, counted in two places), while the hooks in `Recover` would register in `AccordMetrics` but not an `AccordClientRequestMetrics` instance.

@dcapwell Do you have any preference in terms of how we evolve this?

We could leave `AccordClientRequestMetrics` intact, remove any bits from `AccordMetrics` that duplicate those, and revert to `AccordMetrics` (or some other specific metrics class) only tracking timeouts/preempts/etc. related to recovery. We could also leave things as they are, but then we don't have any idea what the breakdown is in `AccordMetrics` between recovery and client requests.

Reply:
> What about the events at the recovery coordinator?

This gets tricky as every node will try to recover from time to time, and it won't know if it actually did anything in some cases[1].  We also get into the situation where multi key txn runs and could double count as we see the events in multiple `CommandStore`, so not sure how you would handle dedup in these cases (or do we just double count?).  I would need to look at this patch, but even having metrics in `Recover` is tricky as our callback doesn't mean we recovered, only that we scheduled things[1]...  The `TransactionStatement` is simple as its logically contained, but that isn't true with recover, so we would need to be careful...

Now, I guess my main question is "how would an operator use these metrics"?  If the answer is "I don't know" then prob best to punt or try to flesh that out.  


[1] `accord.coordinate.Recover#recover()`

```
Deps committedDeps = tryMergeCommittedDeps();
                    node.withEpoch(executeAt.epoch(), () -> {
                        // TODO (required, consider): when writes/result are partially replicated, need to confirm we have quorum of these
                        if (committedDeps != null) Persist.persistMaximal(node, txnId, route, txn, executeAt, committedDeps, acceptOrCommit.writes, acceptOrCommit.result);
                        else CollectDeps.withDeps(node, txnId, route, txn.keys(), executeAt, (deps, fail) -> {
                            if (fail != null) accept(null, fail);
                            else Persist.persistMaximal(node, txnId, route, txn, executeAt, deps, acceptOrCommit.writes, acceptOrCommit.result);
                        });
                    });
                    accept(acceptOrCommit.result, null);
```

Here we trigger the callback saying success, but later on it might actually fail, which will then require another repair to try to push things along...

Reply:
Perhaps, for now, we should just leave `AccordClientRequestMetrics` as it is and remove the overlapping bits in `AccordMetrics`?

Reply:
Is the statement here equally valid for timeouts, preempts and invalidations?

Reply:
I think so. We can always add more metrics around recovery later (especially if they're not going to be well-understood).

Reply:
I hesitate - preemption is something mostly tied to the recovery - isn't it? Let me at least rename those metrics to recoveryTimeouts, ... , and report only what happens in the recover to them? At the same time, such metrics present in `AccordClientRequestMetrics`  report just client facing events?

Reply:
That seems to apply to slow/fast path metrics as well...

talked to @belliottsmith about that - his point is to collect combined and recovery only stats, so yet another idea...

(sorry, previously this comment was accidently added to a wrong thread)

Reply:
My point was only that if we are thinking of separating the stats we should make it easy to view a combined stat since I think this is likely to be the most useful single set of numbers. Separate initial and recovery coordinator stats aren’t independently as useful as a combined stat, though I can see value in wanting to separate them on occasion.

I don’t think the multi key issue that has been raised is anything to worry about - only one key can become coordinator and anyway there’s the same number of coordinators.

statistics we gather per key will be duplicated but presumably these are expected to be duplicated and aren’t coordination level stats.

Reply:
If the most useful single thing we can do is expose combined statistics for both initial and recovery coordinators, why not just keep `AccordMetrics`, remove `AccordClientRequestMetrics`, and proceed from there?

Reply:
agreed

Reply:
@dcapwell regarding your comment, the recovery code updates the metrics in this method:

```java
    @Override
    public void accept(Result result, Throwable failure)
    {
        isDone = true;
        if (failure == null)
        {
            callback.accept(ProgressToken.APPLIED, null);
            node.agent().metricsEventsListener().onRecover(txnId, ballot);
        }
        else
        {
            callback.accept(null, failure);
            if (failure instanceof Preempted)
                node.agent().metricsEventsListener().onPreempted(txnId);
            else if (failure instanceof Timeout)
                node.agent().metricsEventsListener().onTimeout(txnId);
            else if (failure instanceof Invalidated)
                node.agent().metricsEventsListener().onInvalidated(txnId);
        }

        node.agent().onRecover(node, result, failure);
    }
```

As far as I understand, the recovery initiated by this coordinator ends in that method. I don't think we need to deal with deduplication on the coordinator side because the coordinator operates on whole transactions rather than shards. It does not use command stores at all (in fact, it uses them just as `AgentExecutor`, a separate story)